### PR TITLE
Added docs and opinfo for narrow_copy

### PR DESCRIPTION
**Pull request by srossross**

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -12004,6 +12004,17 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
...
            sample_inputs_func=sample_inputs_narrow),
+    OpInfo('narrow_copy',
+           dtypes=all_types_and_complex_and(torch.bool, torch.bfloat16, torch.float16, torch.chalf),
+           supports_out=True,
+           supports_forward_ad=False,
+           supports_fwgrad_bwgrad=False,
+           supports_autograd=False,
+           sample_inputs_func=sample_inputs_narrow,
+           skips=(
+               DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_out'),
```
Comment (srossross) [inline torch/testing/_internal/common_methods_invocations.py:12015]: this test fails because of this issue https://github.com/pytorch/pytorch/issues/84577

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -8174,6 +8174,42 @@ def merge_dicts(*dicts):
 """,
 )
 
+add_docstr(
+    torch.narrow_copy,
+    r"""
```
Comment (amjames) [inline torch/_torch_docs.py:8179]: The online documentation has an entry for `narrow_copy` https://pytorch.org/docs/stable/generated/torch.Tensor.narrow_copy.html

I'm not sure what source it is being generated from as there was no docstring before adding this. We probably want to see if @Lezcano knows where that documentation is coming from and how that would interact with addition of a docstring here. 

Also it is probably a good idea to merge the information somehow, specifically mentioning the motivation for the method is sparse tensor support and how the function interacts with the sparse/dense dimension partition for sparse tensors.

**Review by amjames — COMMENTED** [review]
Looks like a good start, we want to resolve the questions regarding documentation, but the opinfo looks good to me.

File: torch/_torch_docs.py (hunk 2 above)
Comment (Lezcano) [reply torch/_torch_docs.py:8179]: Note that the current documentation is docs for a `Tensor` method. In any case, we can grep a bit of the text as `git grep "except returning a copy rathe"` to find that it lives in `torch/_tensor_docs.py`.

When we have a function that is both a method and a function (as it happens with 99% of operators in PyTorch) we often put the docs in the function, and write a short pointer to the function docs in the method. See for example https://pytorch.org/docs/stable/generated/torch.Tensor.deg2rad.html

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -8174,6 +8174,45 @@ def merge_dicts(*dicts):
...
+    tensor([[ 1,  2,  3],
+            [ 4,  5,  6]])
+    >>> torch.narrow_copy(x, 1, 1, 2)
+    tensor([[ 2,  3],
+            [ 5,  6],
+            [ 8,  9]])
+
+.. seealso::
+
+        :func:`torch.Tensor.narrow_copy` for the method variant
```
Comment (amjames) [inline torch/_torch_docs.py:8210]: The ref to the method variant is not very useful as that now has a signature and link back here.

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -8174,6 +8174,45 @@ def merge_dicts(*dicts):
...
+add_docstr(
+    torch.narrow_copy,
+    r"""
+narrow_copy(input, dim, start, length, *, out=None) -> Tensor
+
+Same as :meth:`Tensor.narrow` except returning a copy rather
+than shared storage.  This is primarily for sparse tensors, which
+do not have a shared-storage narrow method.  Calling ``narrow_copy``
+with ``dimemsion > self.sparse_dim()`` will return a copy with the
+relevant dense dimension narrowed, and ``self.shape`` updated accordingly.
```
Comment (amjames) [inline torch/_torch_docs.py:8186]: I know this text was brought over from the existing docs, but since we are updating it:


The last sentence here seems to imply that one can only apply `narrow_copy` to dense dimensions when the input is sparse. This is not true,  as sparse dims may also be narrowed this way.

Ex:

```python
>>> import torch
>>> s = torch.ones(2,2,2,2).to_sparse(2)
>>> s
tensor(indices=tensor([[0, 0, 1, 1],
                       [0, 1, 0, 1]]),
       values=tensor([[[1., 1.],
                       [1., 1.]],

                      [[1., 1.],
                       [1., 1.]],

                      [[1., 1.],
                       [1., 1.]],

                      [[1., 1.],
                       [1., 1.]]]),
       size=(2, 2, 2, 2), nnz=4, layout=torch.sparse_coo)
>>> s.narrow_copy(0, 0, 1)
tensor(indices=tensor([[0, 0],
                       [0, 1]]),
       values=tensor([[[1., 1.],
                       [1., 1.]],

                      [[1., 1.],
                       [1., 1.]]]),
       size=(1, 2, 2, 2), nnz=2, layout=torch.sparse_coo)
>>> s.narrow_copy(3, 0, 1)
tensor(indices=tensor([[0, 0],
                       [0, 1]]),
       values=tensor([[[1., 1.],
                       [1., 1.]],

                      [[1., 1.],
                       [1., 1.]]]),
       size=(1, 2, 2, 2), nnz=2, layout=torch.sparse_coo)

```

**Review by amjames — CHANGES_REQUESTED** [review]
The opinfo looks good, requesting minor changes to the docs though.

File: torch/_torch_docs.py (hunk 2 above)
Comment (srossross) [reply torch/_torch_docs.py:8179]: updated

File: torch/_torch_docs.py (hunk 3 above)
Comment (srossross) [reply torch/_torch_docs.py]: removed

File: torch/_torch_docs.py (hunk 4 above)
Comment (srossross) [reply torch/_torch_docs.py]: I've removed the reference to dense and added a sparse example

File: torch/_torch_docs.py [hunk 5]
```diff
@@ -8174,6 +8174,54 @@ def merge_dicts(*dicts):
...
 
+add_docstr(
+    torch.narrow_copy,
+    r"""
+narrow_copy(input, dim, start, length, *, out=None) -> Tensor
+
+Same as :meth:`Tensor.narrow` except returning a copy rather
+than shared storage.  This is primarily for sparse tensors, which
+do not have a shared-storage narrow method.  Calling ``narrow_copy``
+with ``dimemsion > self.sparse_dim()`` will return a copy with the
```
Comment (amjames) [inline torch/_torch_docs.py:8185]: This restriction does not actually in the example below you narrow with `dimension=0` and `self.sparse_dim() = 2`.

File: torch/_torch_docs.py [hunk 6]
```diff
@@ -8174,6 +8174,54 @@ def merge_dicts(*dicts):
...
+relevant dimension narrowed, and ``self.shape`` updated accordingly.
+
+Args:
+    input (Tensor): the tensor to narrow
+    dim (int): the dimension along which to narrow
+    start (int): the starting dimension
```
Comment (amjames) [inline torch/_torch_docs.py:8191]: This should probably be "starting offset" as the `dim` argument is the only dimension "starting dimension" is not correct.

**Review by amjames — COMMENTED** [review]
Couple of final suggestions for the docs, the opinfo looks good!

File: torch/_torch_docs.py (hunk 6 above)
Comment (srossross) [reply torch/_torch_docs.py:8191]: This was copied from `narrow` I'll update that too.

File: torch/_torch_docs.py (hunk 5 above)
Comment (srossross) [reply torch/_torch_docs.py:8185]: I'm not sure what this means, This was moved from `torch/_tensor_docs.py` can you propose a different wording?

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -4273,8 +4273,17 @@ def sample_inputs_narrow(op_info, device, dtype, requires_grad, **kwargs):
...
                              requires_grad=requires_grad)
         yield SampleInput(tensor, args=args)
-        # narrow accepts `start` argument to be a Tensor.
-        yield SampleInput(tensor, args=(args[0], torch.tensor(args[1]), args[2]))
+
+
+def sample_inputs_narrow(op_info, device, dtype, requires_grad, **kwargs):
+    '''
+    sample_inputs_narrow accepts the same inputs as narrow_copy, in addition
+    narrow also accepts `start` argument to be a Tensor.
```
Comment (amjames) [inline torch/testing/_internal/common_methods_invocations.py:4281]:
```suggestion
    narrow has an additional overload accepting a Tensor as the start argument.
```

**Review by amjames — APPROVED** [review]
One nit, but otherwise awesome!

**Review by ngimel — APPROVED** [review]

File: torch/_torch_docs.py [hunk 8]
```diff
@@ -8088,6 +8088,52 @@ def merge_dicts(*dicts):
 """,
 )
 
+add_docstr(
+    torch.narrow_copy,
+    r"""
+narrow_copy(input, dim, start, length, *, out=None) -> Tensor
+
+Same as :meth:`Tensor.narrow` except returning a copy rather
```
Comment (mruberry) [inline torch/_torch_docs.py:8096]: "except this returns"

File: torch/testing/_internal/common_methods_invocations.py [hunk 9]
```diff
@@ -4273,8 +4273,16 @@ def sample_inputs_narrow(op_info, device, dtype, requires_grad, **kwargs):
         tensor = make_tensor(shape, dtype=dtype, device=device, low=None, high=None,
                              requires_grad=requires_grad)
         yield SampleInput(tensor, args=args)
-        # narrow accepts `start` argument to be a Tensor.
-        yield SampleInput(tensor, args=(args[0], torch.tensor(args[1]), args[2]))
+
+
+def sample_inputs_narrow(op_info, device, dtype, requires_grad, **kwargs):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:4278]: Nice reuse

File: torch/testing/_internal/common_methods_invocations.py [hunk 10]
```diff
@@ -12018,6 +12026,22 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
                DecorateInfo(unittest.expectedFailure, 'TestMathBits', 'test_conj_view'),
                DecorateInfo(unittest.expectedFailure, 'TestMathBits', 'test_neg_view'),
            )),
+    OpInfo('narrow_copy',
+           dtypes=all_types_and_complex_and(torch.bool, torch.bfloat16, torch.float16, torch.chalf),
+           supports_out=True,
+           supports_forward_ad=False,
+           supports_fwgrad_bwgrad=False,
+           supports_autograd=False,
+           sample_inputs_func=sample_inputs_narrow_copy,
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:12035]: Future TODO: add sparse sample inputs (since this function is intended for use with sparse tensors)

**Review by mruberry — APPROVED** [review]
LGTM! See one small doc tweak suggestion inline

File: torch/_torch_docs.py (hunk 8 above)
Comment (srossross) [reply torch/_torch_docs.py]: updated
